From: Craig Rodrigues <rodrigc@crodrigues.org>
Date: Sat, 20 Jun 2020 03:10:22 -0700
Subject: Fix imap4-utf-7 codec lookup function for Python 3.9

Python 3.9 normalizes the codec name into 'imap4_utf_7' rather than
'imap4-utf-7', and therefore the lookup function needs to account
for the former name.  Transform the latter locally to preserve support
for all Python versions.

Submitted by:  Michal Gorny <mgorny@gentoo.org>

Author: Michal Gorny <mgorny@gentoo.org>
Origin: upstream, https://github.com/twisted/twisted/commit/b05a5b57cd07af2fe6899b7aefa5a42737919419
Last-Updated: 2021-02-13
---
 src/twisted/mail/imap4.py                | 13 ++++++++++++-
 src/twisted/mail/newsfragments/9832.misc |  0
 2 files changed, 12 insertions(+), 1 deletion(-)
 create mode 100644 src/twisted/mail/newsfragments/9832.misc

diff --git a/src/twisted/mail/imap4.py b/src/twisted/mail/imap4.py
index 7949ef5..08ac7cb 100644
--- a/src/twisted/mail/imap4.py
+++ b/src/twisted/mail/imap4.py
@@ -6364,13 +6364,24 @@ class StreamWriter(codecs.StreamWriter):
         return encoder(s)
 
 
+
 _codecInfo = codecs.CodecInfo(encoder, decoder, StreamReader, StreamWriter)
 
 
+
 def imap4_utf_7(name):
-    if name == 'imap4-utf-7':
+    # In Python 3.9, codecs.lookup() was changed to normalize the codec name
+    # in the same way as encodings.normalize_encoding().  The docstring
+    # for encodings.normalize_encoding() describes how the codec name is
+    # normalized.  We need to replace '-' with '_' to be compatible with
+    # older Python versions.
+    #  See:  https://bugs.python.org/issue37751
+    #        https://github.com/python/cpython/pull/17997
+    if name.replace('-', '_') == 'imap4_utf_7':
         return _codecInfo
 
+
+
 codecs.register(imap4_utf_7)
 
 __all__ = [
diff --git a/src/twisted/mail/newsfragments/9832.misc b/src/twisted/mail/newsfragments/9832.misc
new file mode 100644
index 0000000..e69de29
